### ISSUE TITLE
This is not a bug issue!

### ISSUE THREAD

**Issue by jay2013** [stefan-zobel/streamsupport#1]
Hi, 
I am curious about your code implementation. I notice that the parameter type of the constructor of class "java8.util.DelegatingSpliterator" is "java.util.Spliterator" which is only supported by jdk8+.
Why I can use this when my project is compiled by jdk6 or jdk7?

Also, when I de-compile file "streamsupport-1.6.1.jar" you release, we find there are no java.util.Spliterator.class file in this jar file, so I am quite interested in the principle how program run this codes when it is on jdk7 vm?

waiting for your explanation. thanks very much.

best regards,

Dongjie He

**Comment by stefan-zobel** [comment]
It's quite simple. The decision whether we can delegate to a host implementation or not is made at runtime. When the platform has a `JSR 335` implementation we delegate, otherwise we don't.

`streamsupport` is first compiled to bytecode on Java 8 (which has the JSR 335 API) and then rewritten to Java 6 bytecode by retrolambda. That second step rewrites all lambdas and method references into anonymous classes and changes the class version to 50.0 but doesn't check or touch any of the API calls.

You **_can_** have references to non-existing API in bytecode as long as you can guarantee that it never gets loaded at runtime. This behavior is not specified by the VM spec, but it works in practice on Android an on all VMs (Sun, Oracle, IBM) I've tested.

In short: the `java.util.Spliterator` interface never gets loaded / used on a Java 6/7 VM or on Android API level < 24.

Please let me know if you have any follow-up questions.

Regards,
Stefan

**Comment by jay2013** [comment]
Your explanation is very helpful to us. Thank you very much.
